xbrowse/utils/basic_utils.py

```diff
@@ -200,10 +199,80 @@ def slugify(s, separator='_', replace_dot=False):
         string with all characters except [a-Z\-_.] replaced with '_'
     """
     try:
-        regexp = '[^a-zA-Z0-9\-_.]+' if not replace_dot else '[^a-zA-Z0-9\-_]+' 
+        regexp = '[^a-zA-Z0-9\-_.]+' if not replace_dot else '[^a-zA-Z0-9\-_]+'
         words = re.split(regexp, s)
     except Exception as e:
         print("ERROR: string '%s' caused: %s" % (e, s))
-        raise 
+        raise
 
     return separator.join(filter(None, words))
+
+
+# make encoded values as human-readable as possible
+ES_FIELD_NAME_ESCAPE_CHAR = '$'
+ES_FIELD_NAME_BAD_LEADING_CHARS = set(['_', '-', '+', ES_FIELD_NAME_ESCAPE_CHAR])
+ES_FIELD_NAME_SPECIAL_CHAR_MAP = {
+    '.': '_$dot$_',
+    ',': '_$comma$_',
+    '#': '_$hash$_',
+    '*': '_$star$_',
+    '(': '_$lp$_',
+    ')': '_$rp$_',
+    '[': '_$lsb$_',
+    ']': '_$rsb$_',
+    '{': '_$lcb$_',
+    '}': '_$rcb$_',
+}
+
+
+def _encode_name(s):
+    """Applies a reversable encoding to the special chars in the given name or id string, and returns the result.
+    Among other things, the encoded string is a valid elasticsearch or mongodb field name.
+
+    See:
+    https://discuss.elastic.co/t/special-characters-in-field-names/10658/2
+    https://discuss.elastic.co/t/illegal-characters-in-elasticsearch-field-names/17196/2
+    """
+    field_name = StringIO.StringIO()
+    for c in s:
+        if c == ES_FIELD_NAME_ESCAPE_CHAR:
+            field_name.write(2*ES_FIELD_NAME_ESCAPE_CHAR)
+        elif c in ES_FIELD_NAME_SPECIAL_CHAR_MAP:
+            field_name.write(ES_FIELD_NAME_SPECIAL_CHAR_MAP[c])  # encode the char
+        else:
+            field_name.write(c)  # write out the char as is
+
+    field_name = field_name.getvalue()
+
+    # escape 1st char if necessary
+    if any(field_name.startswith(c) for c in ES_FIELD_NAME_BAD_LEADING_CHARS):
+        return ES_FIELD_NAME_ESCAPE_CHAR + field_name
+    else:
+        return field_name
+
+
+def _decode_name(s):
+    """Decodes a name or id string that was encoded by #_encode_name(..) and returns the original string"""
+
+    if s.startswith(ES_FIELD_NAME_ESCAPE_CHAR):
+        s = s[1:]
+
+    i = 0
+    original_string = StringIO.StringIO()
+    while i < len(s):
+        current_string = s[i:]
+        if current_string.startswith(2*ES_FIELD_NAME_ESCAPE_CHAR):
```

Comment:
what if you escaped one of the other bad start characters?

Reply:
hmm - special chars at the beginning of the string - yeah, I didn't have tests for that, so just added a couple.  These are the cases you were thinking of, right? 

```
   ...: for test_string in [
   ...:             ".SPECIAL_START_CHAR",
   ...:             "*SPECIAL_START_CHAR",
   ...:             "__SPECIAL_START_CHAR",
   ...:             "__$SPECIAL_START_CHAR",
   ...:             "_$SPECIAL_START_CHAR",
   ...:         ]:
   ...:             encoded_test_string = _encode_name(test_string)
   ...:             decoded_test_string = _decode_name(encoded_test_string)
   ...:             print("original: "+ test_string)
   ...:             print("encoded: " + encoded_test_string)
   ...:             print("decoded: " + decoded_test_string)
   ...:             print("----")
   ...:
   ...:
   ...:
original: .SPECIAL_START_CHAR
encoded: $_$dot$_SPECIAL_START_CHAR
decoded: .SPECIAL_START_CHAR
----
original: *SPECIAL_START_CHAR
encoded: $_$star$_SPECIAL_START_CHAR
decoded: *SPECIAL_START_CHAR
----
original: __SPECIAL_START_CHAR
encoded: $__SPECIAL_START_CHAR
decoded: __SPECIAL_START_CHAR
----
original: __$SPECIAL_START_CHAR
encoded: $__$$SPECIAL_START_CHAR
decoded: __$SPECIAL_START_CHAR
----
original: _$SPECIAL_START_CHAR
encoded: $_$$SPECIAL_START_CHAR
decoded: _$SPECIAL_START_CHAR
----
```

Reply:
yup, thats what I was thinking